### PR TITLE
Tickets/dm 30898

### PR DESCRIPTION
**Pull request by wvreeven**
Creating a draft PR for now.

### CODE REVIEW

File: python/lsst/ts/dimm/controllers/base_dimm.py
Comment (tribeiro) [inline python/lsst/ts/dimm/controllers/base_dimm.py]: Uhm, why do you need this new flag here? From the code bellow it seems like `controller_running = False` is equivalent to `status = DIMMStatus["INITIALIZED"]` and `controller_running = True` is equivalent to `status = DIMMStatus["RUNNING"]` .

I did just realized a problem in the implementation of the status flag. The original idea was that this would be a bitmap so we could flip bits on and off to indicate each status. 

For instance; you could have the status as INITIALIZED and RUNNING: `1<<1 | 1<<2`

and also INITIALIZED and RUNNING and ERROR: `1<<1 | 1<<2 | 1<<3`

But I just noticed that the implementation is not carrying it out like this. 

My suggestion is that you remove this flag and, for now, assume `controller_running = status ==  DIMMStatus["RUNNING"]`. Then later we fix the bitmap operation and, if you still think this status is needed, we add a new "CONTROLLERRUNNING" bit...

File: python/lsst/ts/dimm/dimm_csc.py
Comment (tribeiro) [inline python/lsst/ts/dimm/dimm_csc.py:211]: Since you are not raising the exception, the command will actually succeed. Worst than that, the CSC will proceed to execute the next line, starting the telemetry loop. Consider doing the following instead:

```
         try:
             await self.controller.start()
         except Exception:
             self.fault(
                   code=CONTROLLER_START_FAILED,
                   report="Failed to start controller after transitioning to enable state.",
                   traceback=traceback.format_exc(),
              )
             raise RuntimeError("Failed to start controller. Check configuration and make sure DIMM controller is alive and reachable by the CSC.")
```

This will make sure to reject and interrupt the command execution, and will give users a feedback on the error.

File: python/lsst/ts/dimm/dimm_csc.py
Comment (tribeiro) [inline python/lsst/ts/dimm/dimm_csc.py]: The CSC class seems like an odd place to have this method. Maybe create a `utils` submodule?

I think `convert_to_int` might be a better name? `convert_int` seems like you are converting and int to something... 

Also, are you sure 0 is the fallback value you want? what is the sentinel values we agreed with SQuaRE?

File: python/lsst/ts/dimm/dimm_csc.py
Comment (tribeiro) [inline python/lsst/ts/dimm/dimm_csc.py]: same comments here as for `convert_int`. :)

File: python/lsst/ts/dimm/dimm_csc.py
Comment (tribeiro) [inline python/lsst/ts/dimm/dimm_csc.py]: so, this should probably be renamed `convert_dimm_measurement_data`. Mixing camelCase with snake_case is not a good idea.. 

The fact that the only class method you are calling here in `convert_float` (which I think don't belong in the class), tells me that this method should probably also be part of the same utils module as `convert_float` (`convert_to_float`) and `convert_int` (`convert_to_int`)..

Also, consider refactoring this method to make it more "dynamic". It will be easier to add/remove new fields later if you do something like this:

```
        return dict(
            [
                (key, convert_to_float(data[key]))
                for key in {
                    "timestamp",
                    "hrNum",
                    "secz",
                    "fwhm",
                    "fwhmx",
                    "fwhmy",
                    "r0",
                    "nimg",
                    "dx",
                    "dy",
                    "flux",
                    "fluxL",
                    "scintL",
                    "strehlL",
                    "fluxR",
                    "scintR",
                    "strehlR",
                }
            ]
        )
```

File: python/lsst/ts/dimm/controllers/base_dimm.py
Comment (wvreeven) [reply python/lsst/ts/dimm/controllers/base_dimm.py]: I introduced the new flag to be able to query the DIMM server status. DIMMStatus["RUNNING"] is set already when the CSC connects to the DIMM hardware, which doesn't automatically mean that it actually is producing sensible data. Judging by that, the `controller_running = status == DIMMStatus["RUNNING"]` doesn't seem to add much to the existing code. But OK, I followed your suggestion.

File: python/lsst/ts/dimm/dimm_csc.py
Comment (wvreeven) [reply python/lsst/ts/dimm/dimm_csc.py:211]: Thanks for catching that. Done.

File: python/lsst/ts/dimm/dimm_csc.py
Comment (wvreeven) [reply python/lsst/ts/dimm/dimm_csc.py]: Agreed. There is only an agreed value for floats (which is math.nan). There is no default value for ints. I am open to suggestions :)

File: python/lsst/ts/dimm/dimm_csc.py
Comment (wvreeven) [reply python/lsst/ts/dimm/dimm_csc.py]: Agreed.

**Review by tribeiro — COMMENTED** [review]
Left a few comments in line...

File: python/lsst/ts/dimm/controllers/base_dimm.py
Comment (tribeiro) [reply python/lsst/ts/dimm/controllers/base_dimm.py]: If you have a reason for having `controller_running` you should probably leave it then... and we update it when refactoring the use of the status bit mask.

File: python/lsst/ts/dimm/dimm_csc.py
Comment (tribeiro) [reply python/lsst/ts/dimm/dimm_csc.py]: probably the same as for float...

File: python/lsst/ts/dimm/dimm_csc.py
Comment (wvreeven) [reply python/lsst/ts/dimm/dimm_csc.py]: Done, though slightly different than you suggested because some values need to be converted to ints.

File: python/lsst/ts/dimm/controllers/base_dimm.py
Comment (wvreeven) [reply python/lsst/ts/dimm/controllers/base_dimm.py]: Actually I think we can use the DIMMStatus already, but the point at which is gets set to `RUNNING` should be changed. I'll implement that and will push the code so you can see if you agree with me.

File: python/lsst/ts/dimm/controllers/astelco_dimm.py
Comment (tribeiro) [inline python/lsst/ts/dimm/controllers/astelco_dimm.py]: Not sure I understand this comment. the f-string will deal with whatever type the input is. Consider removing it

File: python/lsst/ts/dimm/controllers/base_dimm.py
Comment (tribeiro) [inline python/lsst/ts/dimm/controllers/base_dimm.py]: This is not exactly the status of the CSC but of the controller. I think we can get by without this comment here. If you feel it is necessary consider changing it to:

> Status of the DIMM controller.

**Review by tribeiro — COMMENTED** [review]
`python/lsst/ts/dimm/utils/__init__.py` is empty. Do you want to add 

```
from .conversion import *
```

?

File: python/lsst/ts/dimm/utils/conversion.py
Comment (tribeiro) [inline python/lsst/ts/dimm/utils/conversion.py:21]: add:

```
__all__ = ["convert_to_int", "convert_to_float", "convert_dimm_measurement_data"]
```

File: python/lsst/ts/dimm/utils/conversion.py
Comment (tribeiro) [inline python/lsst/ts/dimm/utils/conversion.py:44]: should be `math.nan`

**Review by tribeiro — APPROVED** [review]
Left a few additional comments. Please, take a look before merging. Also, please, rebase/squash the commits into an appropriate number of commits. I'd recommend trying to squash them such that you get 1 feature per commit, but it may be easier to squash them all into 1 commit. I'll leave that to your best judgement. :)

File: python/lsst/ts/dimm/controllers/astelco_dimm.py
Comment (wvreeven) [reply python/lsst/ts/dimm/controllers/astelco_dimm.py]: Done

File: python/lsst/ts/dimm/controllers/base_dimm.py
Comment (wvreeven) [reply python/lsst/ts/dimm/controllers/base_dimm.py]: DIMMStatus gets publicly exported so I'd like to keep a comment for it. I changed it according to your suggestion.

File: python/lsst/ts/dimm/utils/conversion.py
Comment (wvreeven) [reply python/lsst/ts/dimm/utils/conversion.py:21]: Both done.

File: python/lsst/ts/dimm/utils/conversion.py
Comment (wvreeven) [reply python/lsst/ts/dimm/utils/conversion.py:44]: We discussed this on slack and agreed to leave it to `0` since `math.nan` is a float and there is no preferred sentinel value for integers.

**Pull request merged**
